pylmnn/impostor_sampling.py

```diff
@@ -0,0 +1,164 @@
+        latest_values : array-like, shape (n_samples,)
+            The latest part of the stream to sample from.
+
+        """
+        self._values.extend(latest_values)
+
+    def sample(self):
+        """Gets the current sample.
+
+        Returns
+        -------
+        sample : array, shape (sample_size,)
+            The current sample from the stream
+
+        """
+        n = len(self._values)
+        k = self.sample_size
+        if n > k:
+            ind = self.random_state.choice(n, k, replace=False)
+            return np.asarray(self._values, dtype=object)[ind]
+        else:
+            return self._values
+
+
+class ReservoirSampler(Sampler):
+    """Pseudo-uniform online sample of fixed size from a stream of values.
+
+    Parameters
+    ----------
+    sample_size : int
+        Number of elements to take as the sample.
+
+    random_state : numpy.RandomState
+        A pseudo random number generator object used for sampling.
+
+    References
+    ----------
+
+    .. [1] Li, Kim-Hung (4 December 1994).
+           "Reservoir-Sampling Algorithms of Time Complexity O(n(1+log(N/n)))"
+           ACM Transactions on Mathematical Software. 20 (4): 481–493.
+    """
+
+    def __init__(self, sample_size, random_state):
+        self.sample_size = sample_size
+        self.random_state = random_state
+
+        self._i = 0
+        self._reservoir = []
+
+        u = random_state.random()
+        self._w = np.exp(np.log(u) / sample_size)
+
+        u = random_state.random()
+        self._next_i = int(sample_size + np.log(u) / np.log(1 - self._w) + 1)
+
+        # stores randomly generated values for sampling which we do in batches
+        # to minimize overhead of getting random numbers
+        # precomputing and  storing these per-sampled-element values is over
+        # twice as fast as naively generating the random numbers and computing
+        # them inside the sampling loop
+        self._rand_replacement = None
+        self._rand_w_multiplier = None
+        self._rand_next_i_partial = None
+        self._rand_ind = sample_size
+
+    def extend(self, latest_values):
+        """Adds elements to the reservoir sample.
+
+        Parameters
+        ----------
+        latest_values : array-like, shape (n_samples,)
+            The latest part of the stream to sample from.
+
+        """
+
+        offset = self._i
+        k = self.sample_size
+        rng = self.random_state
+
+        latest_values = list(latest_values)  # TODO: This seems unnecessary
+        remaining_capacity = k - offset
+
+        # if reservoir not full, copy from latest_values to fill
+        if remaining_capacity > 0:
+            to_copy = min(remaining_capacity, len(latest_values))
+            self._reservoir.extend(latest_values[:to_copy])
+
+        # NOTE: At this point, the reservoir might still be not full
+
+        # general case: sample from latest_values
+        while self._next_i - offset < len(latest_values):
```

Comment:
I think this while loop can be vectorized. I need to look at this in more depth, but I think a lot of time can be saved.

Reply:
yeah, I agree, and this is where things are slower for the reservoir sampler. Related to this is the idea of pre-computing the sample indicies in blocks. I could easily do some of this (https://github.com/johny-c/pylmnn/pull/5/files/42d42ab02d310aa4f5a8d7b716ae2bb50396aa16#diff-adeb34b813966f09fdcba22cd080044cf631ad669d04a94a05f971dd51c868b0R131-R139), but my math of random distributions isn't quite good enough to vectorize all of it.

More concisely: I believe it's possible to create a vector of all the necessary elements from the geometric distribution for the given block we're extending from. I just don't know how.